crates/red_knot_python_semantic/src/types.rs

```diff
@@ -729,26 +675,152 @@ impl<'db> Type<'db> {
                             })
                         })
             }
-            (Type::Intersection(intersection), ty) => intersection
+            (Type::Intersection(intersection), _) => intersection
                 .positive(db)
                 .iter()
-                .any(|&elem_ty| elem_ty.is_subtype_of(db, ty)),
-            (ty, Type::Intersection(intersection)) => {
+                .any(|&elem_ty| elem_ty.is_subtype_of(db, target)),
+            (_, Type::Intersection(intersection)) => {
                 intersection
                     .positive(db)
                     .iter()
-                    .all(|&pos_ty| ty.is_subtype_of(db, pos_ty))
+                    .all(|&pos_ty| self.is_subtype_of(db, pos_ty))
                     && intersection
                         .negative(db)
                         .iter()
-                        .all(|&neg_ty| neg_ty.is_disjoint_from(db, ty))
+                        .all(|&neg_ty| neg_ty.is_disjoint_from(db, self))
```

Comment:
Not related to your PR, but it would look a bit more symmetric if we change the order here to what we have above for the subtype-of check for positive elements:
```suggestion
                        .all(|&neg_ty| self.is_disjoint_from(db, neg_ty))
```